src/praisonai-bot/tests/unit/test_bot_gap_features.py

```
@@ -328,11 +328,27 @@ def test_env_var_resolution(self):
 
-    def test_missing_env_var_fails(self):
+    def test_missing_env_var_isolates_channel(self):
+        # Partial-credential isolation (Issue #3159): an unset channel token
+        # env var must NOT abort the whole gateway config. The channel keeps
+        # a degraded (empty) token so the runtime skips only that channel and
+        # every healthy channel keeps serving.
         from praisonai_bot.bots._config_schema import validate_bot_config
         os.environ.pop("_NONEXISTENT_TOKEN", None)
-        with pytest.raises(ValueError, match="not set"):
-            validate_bot_config({"channels": {"telegram": {"token": "${_NONEXISTENT_TOKEN}"}}})
+        os.environ["_TEST_HEALTHY_TOKEN"] = "healthy-token"
```

Ruff (0.15.21)

[error] Possible hardcoded password assigned to: "_TEST_HEALTHY_TOKEN"

(S105)